core/distributed/vector.cpp

```diff
@@ -0,0 +1,394 @@
+{
+    auto exec = result->get_executor();
+
+    GKO_ASSERT(partition->get_executor() == exec);
+
+    auto global_rows = static_cast<size_type>(partition->get_size());
+    auto global_cols = data.get_size()[1];
+    auto tmp = Vector<ValueType>::create(exec, result->get_communicator(),
+                                         gko::dim<2>{global_rows, global_cols});
+
+    auto rank = tmp->get_communicator().rank();
+    auto local_rows = static_cast<size_type>(partition->get_part_size(rank));
+    if (tmp->get_local()->get_size() != dim<2>{local_rows, global_cols}) {
+        auto stride = tmp->get_local()->get_stride() > 0
+                          ? tmp->get_local()->get_stride()
+                          : global_cols;
+        Vector<ValueType>::local_vector_type::create(
+            exec, dim<2>{local_rows, global_cols}, stride)
+            ->move_to(tmp->get_local());
+    }
+    tmp->get_local()->fill(zero<ValueType>());
+    exec->run(
+        vector::make_build_local(data, partition, rank, tmp->get_local()));
+    tmp->move_to(result);
+}
+
+#define GKO_DECLARE_DISTRIBUTED_READ_DISTRIBUTED_IMPL(               \
+    ValueType, LocalIndexType, GlobalIndexType)                      \
+    void read_distributed_impl(                                      \
+        const device_matrix_data<ValueType, GlobalIndexType>& data,  \
+        const Partition<LocalIndexType, GlobalIndexType>* partition, \
+        Vector<ValueType>* storage)
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_LOCAL_GLOBAL_INDEX_TYPE(
+    GKO_DECLARE_DISTRIBUTED_READ_DISTRIBUTED_IMPL);
+
+
+}  // namespace detail
+
+
+template <typename ValueType>
+void Vector<ValueType>::apply_impl(const LinOp* b, LinOp* x) const
+{
+    GKO_NOT_SUPPORTED(this);
+}
+
+
+template <typename ValueType>
+void Vector<ValueType>::apply_impl(const LinOp* alpha, const LinOp* b,
+                                   const LinOp* beta, LinOp* x) const
+{
+    GKO_NOT_SUPPORTED(this);
+}
+
+template <typename ValueType>
+Vector<ValueType>::Vector(std::shared_ptr<const Executor> exec,
+                          mpi::communicator comm, dim<2> global_size,
+                          dim<2> local_size)
+    : Vector(exec, comm, global_size, local_size, local_size[1])
+{}
+
+template <typename ValueType>
+Vector<ValueType>::Vector(std::shared_ptr<const Executor> exec,
+                          mpi::communicator comm, dim<2> global_size,
+                          dim<2> local_size, size_type stride)
+    : EnableLinOp<Vector<ValueType>>{exec, global_size},
+      DistributedBase{comm},
+      local_{exec, local_size, stride}
+{}
+
+
+template <typename ValueType>
+void Vector<ValueType>::fill(const ValueType value)
+{
+    this->get_local()->fill(value);
+}
+
+
+template <typename ValueType>
+void Vector<ValueType>::convert_to(
+    Vector<next_precision<ValueType>>* result) const
+{
+    result->set_size(this->get_size());
+    result->set_communicator(this->get_communicator());
```

Comment:
Are communicators immutable properties like executors or mutable? I would prefer the former, and fail if the communicators have different sizes.